x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -0,0 +1,160 @@
+ */
+public final class MlAutoscalingResourceTracker {
+
+    private MlAutoscalingResourceTracker() {}
+
+    public static void getMlAutoscalingStats(
+        ClusterState clusterState,
+        Client client,
+        TimeValue timeout,
+        MlMemoryTracker mlMemoryTracker,
+        ActionListener<MlAutoscalingStats> listener
+    ) {
+        String[] mlNodes = clusterState.nodes()
+            .stream()
+            .filter(node -> node.getRoles().contains(DiscoveryNodeRole.ML_ROLE))
+            .map(DiscoveryNode::getId)
+            .toArray(String[]::new);
+
+        getMlNodeStats(
+            mlNodes,
+            client,
+            timeout,
+            ActionListener.wrap(
+                osStatsPerNode -> getMemoryAndCpu(clusterState, mlMemoryTracker, osStatsPerNode, listener),
+                listener::onFailure
+            )
+        );
+
+    }
+
+    static void getMlNodeStats(String[] mlNodes, Client client, TimeValue timeout, ActionListener<Map<String, OsStats>> listener) {
+        client.admin()
+            .cluster()
+            .prepareNodesStats(mlNodes)
+            .clear()
+            .setOs(true)
+            .setTimeout(timeout)
+            .execute(
+                ActionListener.wrap(
+                    nodesStatsResponse -> listener.onResponse(
+                        nodesStatsResponse.getNodes()
+                            .stream()
+                            .collect(Collectors.toMap(nodeStats -> nodeStats.getNode().getId(), NodeStats::getOs))
+                    ),
+                    listener::onFailure
+                )
+            );
+    }
+
+    static void getMemoryAndCpu(
+        ClusterState clusterState,
+        MlMemoryTracker mlMemoryTracker,
+        Map<String, OsStats> osStatsPerNode,
+        ActionListener<MlAutoscalingStats> listener
+    ) {
+        Set<String> nodesWithRunningJobs = new HashSet<>();
+        long memoryBytesSum = osStatsPerNode.values()
+            .stream()
+            .map(s -> s.getMem().getAdjustedTotal().getBytes())
+            .collect(Collectors.summingLong(Long::longValue));
+
+        long modelMemoryBytesSum = 0;
+        long extraSingleNodeModelMemoryInBytes = 0;
+        int extraSingleNodeProcessors = 0;
+        int minNodes = 0;
+        int extraProcessors = 0;
+
+        final MlAutoscalingContext autoscalingContext = new MlAutoscalingContext(clusterState);
+
+        for (var task : autoscalingContext.anomalyDetectionTasks) {
+            String jobId = ((OpenJobAction.JobParams) task.getParams()).getJobId();
+            if (AWAITING_LAZY_ASSIGNMENT.equals(task.getAssignment())) {
+                extraSingleNodeModelMemoryInBytes = Math.max(
+                    extraSingleNodeModelMemoryInBytes,
+                    mlMemoryTracker.getAnomalyDetectorJobMemoryRequirement(jobId)
+                );
+                extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, 1);
+                ++extraProcessors;
```

Comment:
Currently DFA doesn't require any CPU either. The current logic for processors is in: https://github.com/elastic/elasticsearch/blob/d704f21ef9a54ee504221879a47c98b628383d5a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlProcessorAutoscalingDecider.java#L114-L137

Currently that only considers trained models.

(I don't think it's necessary to carry over the 0.1 processor requirement for low priority models. The memory requirement should ensure low priority models get a node to run on.)

Reply:
It's probably a good idea to add comments in both the new and old autoscaling code to say that if we ever change this decision so that jobs require processors then both places should be changed.